### ISSUE TITLE
include_role: role handler not found

### ISSUE THREAD

**Issue by tchernomax** [ansible/ansible#31071]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

include_role

##### ANSIBLE VERSION

```
ansible 2.5.0
  config file = /home/max/.ansible.cfg
  configured module search path = ['/home/max/.ansible/modules']
  ansible python module location = /home/max/Telechargements/pyvirt/lib/python3.6/site-packages/ansible
  executable location = /home/max/Telechargements/pyvirt/bin/ansible
  python version = 3.6.2 (default, Jul 20 2017, 03:52:27) [GCC 7.1.1 20170630]
```

##### CONFIGURATION

```
ALLOW_WORLD_READABLE_TMPFILES(/home/max/.ansible.cfg) = True
DEFAULT_HOST_LIST(/home/max/.ansible.cfg) = ['/home/max/.ansible/oxalide-hosts']
DEFAULT_MODULE_PATH(/home/max/.ansible.cfg) = ['/home/max/.ansible/modules']
RETRY_FILES_ENABLED(/home/max/.ansible.cfg) = False
```

##### OS / ENVIRONMENT

N/A

##### SUMMARY

handler from a role called via "include_role" module, from another role called via the "roles" array of a play, aren't found if the play contain one or more handlers.

##### STEPS TO REPRODUCE

```
max@testhost % LC_ALL=C tree --noreport
.
|-- playbook.yml
`-- roles
    |-- role1
    |   `-- tasks
    |       `-- main.yml
    `-- role2
        |-- handlers
        |   `-- main.yml
        `-- tasks
            `-- main.yml

max@testhost % tail -n +1 **/*.yml
==> playbook.yml <==
---
- hosts: localhost
  gather_facts: false

  roles:
    - role: role1

  handlers:
    - name: playbook handler
      debug:

==> roles/role1/tasks/main.yml <==
---
- include_role:
    name: role2

==> roles/role2/handlers/main.yml <==
---
- name: role2 handler
  debug:

==> roles/role2/tasks/main.yml <==
---
- name: role2
  debug:
  changed_when: true
  notify: role2 handler
```

##### EXPECTED RESULTS

```
max@testhost % ansible-playbook playbook.yml --inventory localhost, --connection local

PLAY [localhost] *************************************************************************

TASK [role2 : role2] *********************************************************************
ok: [localhost] => {
    "msg": "Hello world!"
}

RUNNING HANDLER [role2 : role2 handler] **************************************************
ok: [localhost] => {
    "msg": "Hello world!"
}

PLAY RECAP *******************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0

```

##### ACTUAL RESULTS

```
max@testhost % ansible-playbook playbook.yml --inventory localhost, --connection local

PLAY [localhost] ******************************************************************************

TASK [role2 : role2] ***************************************************************************
ERROR! The requested handler 'role2 handler' was not found in either the main handlers list nor in the listening handlers list

```

Same with -vvvv:
```
max@testhost % ansible-playbook -vvvv playbook.yml --inventory localhost, --connection local
ansible-playbook 2.5.0
  config file = /home/max/.ansible.cfg
  configured module search path = ['/home/max/.ansible/modules']
  ansible python module location = /home/max/Telechargements/pyvirt/lib/python3.6/site-packages/ansible
  executable location = /home/max/Telechargements/pyvirt/bin/ansible-playbook
  python version = 3.6.2 (default, Jul 20 2017, 03:52:27) [GCC 7.1.1 20170630]
Using /home/max/.ansible.cfg as config file
setting up inventory plugins
Set default localhost to localhost
Parsed localhost, inventory source with host_list plugin
Loading callback plugin default of type stdout, v2.0 from /home/max/Telechargements/pyvirt/lib/python3.6/site-packages/ansible/plugins/callback/__init__.py

PLAYBOOK: playbook.yml *******************************************************************
1 plays in playbook.yml

PLAY [localhost] *************************************************************************
META: ran handlers

TASK [role2 : role2] *********************************************************************
task path: /home/max/Telechargements/test/roles/role2/tasks/main.yml:2
ERROR! The requested handler 'role2 handler' was not found in either the main handlers list nor in the listening handlers list

```

##### NOTES

The handler is found (the bug doesn't show) if:

* the play doesn't have any handler, or
* role1 is called via include_role in the play tasks (instead of the play roles), or
* `static: false` is added at the "include_role role2" task in role1, or
* role2 is called as role1 dependency (instead of via include_role)

handlers of role1 are found.

**Comment by LuckySB** [comment]
very old bug. that is feature
workaround:
```
|-- playbook.yml
`-- roles
    |-- handlers
    |   `-- handlers
    |       `-- main.yml
    `-- role2
    |   |-- handlers
    |   |   `-- main.yml
    |   `-- tasks
    |        `-- main.yml
    `-- role3
        |-- handlers
        |   `-- main.yml
        `-- tasks
            `-- main.yml
```

write all handlers to roles/handlers/handlers/main.yml

write to roles/role(2,3)/handlers/main.yml
```
  - import_tasks: ../../handlers/handlers/main.yml
```

**Comment by tchernomax** [comment]
@LuckySB thank for your reply, but I don't really like your workaround… It make it more difficult to share roles.
As workaround I prefer to set `static: false` on `include_role` tasks inside roles. But static is now [deprecated](https://github.com/ansible/ansible/blob/cd2774af7849be79f19b43628a3e3dde4bde00b4/lib/ansible/playbook/helpers.py#L141).
It's very weird that `static: false` change the ansible behavior since include_role is normally dynamic by default (so `static: false` is redundant).

**Comment by ghost** [comment]
Similar issue here. 

Ansible version: 2.4
Server Operating system: Ubuntu Xenial 16.04
Client Operating system: Arch Linux

**This is the playbook 'test.yml':**
```
- hosts: cluster
  handlers:
    - name: docker rm -f nginx
      shell: docker rm -f nginx
  tasks:
    - include_role:
        name: pdnsd
```

**These are the role's handlers './roles/pdnsd/handlers/main.yml':**
```
- name: pdnsd-ctl empty-cache
  shell: pdnsd-ctl empty-cache

- name: restart pdnsd
  service:
    name: pdnsd
    state: restarted
    enabled: yes
```

**Because 'handlers:' section was declared in the playbook 'test.yml' above, the role 'pdnsd' was unable to find the handlers declared in './roles/pdnsd/handlers/main.yml':**
```
PLAY [cluster] ****************************************************************************************************************************

TASK [pdnsd : Install pdnsd] *********************************************************************************************
skipping: [cluster] => (item=[]) 

TASK [pdnsd : create file /etc/hosts.custom] *****************************************************************************************
ERROR! The requested handler 'restart pdnsd' was not found in either the main handlers list nor in the listening handlers list
```

It looks like the role 'pdnsd' used the playbook's 'handlers:' instead of the role's handlers, which isn't the expected role behavior (a role is supposed to be independent from the playbook).

****This looks like a bug in 'include_role'. IMHO, Ansible should always use the role's handlers, no matter the case.** The handlers section in a playbook should be used by the playbooks' instructions.**

**Comment by tchernomax** [comment]
@Asher256 your issue is fixed in 2.5.

```
max@testhost % tail -n+1 **/*.yml
==> playbook.yml <==
---
- hosts: localhost
  gather_facts: false

  tasks:
    - include_role:
        name: role1

  handlers:
    - name: playbook handler
      debug:

==> roles/role1/handlers/main.yml <==
---
- name: role1 handler
  debug:

==> roles/role1/tasks/main.yml <==
---
- name: role1
  debug:
  changed_when: true
  notify: role1 handler
```

2.4:
```
max@testhost % ansible --version
ansible 2.4.0.0
  config file = /home/max/.ansible.cfg
  configured module search path = [u'/home/max/.ansible/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Sep 20 2017, 01:25:59) [GCC 7.2.0]

max@testhost % ansible-playbook playbook.yml --inventory localhost, --connection local

PLAY [localhost] ******************************************

TASK [role1 : role1] ***************************************
ERROR! The requested handler 'role1 handler' was not found in either the main handlers list nor in the listening handlers list
```

2.5:
```
max@testhost % ansible --version
ansible 2.5.0
  config file = /home/max/.ansible.cfg
  configured module search path = ['/home/max/.ansible/modules']
  ansible python module location = /home/max/Telechargements/pyvirt/lib/python3.6/site-packages/ansible
  executable location = /home/max/Telechargements/pyvirt/bin/ansible
  python version = 3.6.2 (default, Jul 20 2017, 03:52:27) [GCC 7.1.1 20170630]

max@testhost % ansible-playbook playbook.yml --inventory localhost, --connection local

PLAY [localhost] *******************************************************

TASK [role1 : role1] ****************************************************
ok: [localhost] => {
    "msg": "Hello world!"
}

RUNNING HANDLER [role1 : role1 handler] *************************
ok: [localhost] => {
    "msg": "Hello world!"
}

PLAY RECAP **********************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0

```

**Comment by ghost** [comment]
+1 tchernomax

Thank you for the information.

**Comment by ssbarnea** [comment]
@Asher256 do you know which fix sorted this bug in 2.5? I want to get it backported to stable-2.4

**Comment by rpijlman** [comment]
Clone of #18140 and #26718

**Comment by mkrizek** [comment]
@tchernomax This seems to be fixed in https://github.com/ansible/ansible/pull/31463 which has been merged and is included in `2.4.1`. Can you please re-test? Thanks!

needs_info

**Comment by tchernomax** [comment]
@mkrizek indeed, it's now fixed.

```
(venv) max@test % ansible --version
ansible 2.5.0
  config file = /home/max/.ansible.cfg
  configured module search path = ['/home/max/.ansible/modules']
  ansible python module location = /home/max/Telechargements/venv/lib/python3.6/site-packages/ansible
  executable location = /home/max/Telechargements/venv/bin/ansible
  python version = 3.6.4 (default, Dec 23 2017, 19:07:07) [GCC 7.2.1 20171128]

(venv) max@test % ansible-playbook playbook.yml --inventory localhost, --connection local

PLAY [localhost] ******************************

TASK [role1 : include_role] *******************

TASK [role2 : role2] **************************
changed: [localhost] => {
    "attempts": 1,
    "msg": "Hello world!"
}

RUNNING HANDLER [role2 : role2 handler] *******
ok: [localhost] => {
    "attempts": 1,
    "msg": "Hello world!"
}

PLAY RECAP ************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0
```
